noty-android/app/src/main/java/dev/shreyaspatil/noty/view/viewmodel/AddNoteViewModel.kt

```diff
@@ -16,50 +16,66 @@
 import dagger.hilt.android.lifecycle.HiltViewModel
 import dev.shreyaspatil.noty.core.model.NotyTask
 import dev.shreyaspatil.noty.core.repository.NotyNoteRepository
-import dev.shreyaspatil.noty.core.repository.ResponseResult
 import dev.shreyaspatil.noty.core.task.NotyTaskManager
-import dev.shreyaspatil.noty.core.ui.UIDataState
 import dev.shreyaspatil.noty.di.LocalRepository
-import dev.shreyaspatil.noty.utils.ext.shareWhileObserved
-import javax.inject.Inject
+import dev.shreyaspatil.noty.utils.validator.NoteValidator
+import dev.shreyaspatil.noty.view.state.AddNoteState
 import kotlinx.coroutines.Job
-import kotlinx.coroutines.flow.MutableSharedFlow
 import kotlinx.coroutines.launch
+import javax.inject.Inject
 
 @HiltViewModel
 class AddNoteViewModel @Inject constructor(
     @LocalRepository private val noteRepository: NotyNoteRepository,
     private val notyTaskManager: NotyTaskManager
-) : ViewModel() {
+) : BaseViewModel<AddNoteState>(initialState = AddNoteState()) {
+
+    private var job: Job? = null
 
-    var job: Job? = null
+    fun setTitle(title: String) {
+        setState { state -> state.copy(title = title) }
+        validateNote()
+    }
 
-    private val _addNoteState = MutableSharedFlow<UIDataState<String>>()
-    val addNoteState = _addNoteState.shareWhileObserved(viewModelScope, 0)
+    fun setNote(note: String) {
+        setState { state -> state.copy(note = note) }
+        validateNote()
+    }
 
-    fun addNote(title: String, note: String) {
+    fun add() {
         job?.cancel()
         job = viewModelScope.launch {
-            _addNoteState.emit(UIDataState.loading())
+            val title = state.value.title.trim()
+            val note = state.value.note.trim()
+
+            setState { state -> state.copy(isAdding = true) }
 
-            val state = when (val result = noteRepository.addNote(title, note)) {
-                is ResponseResult.Success -> {
-                    val noteId = result.data
-                    scheduleNoteCreate(noteId)
-                    UIDataState.success(noteId)
+            val result = noteRepository.addNote(title, note)
+
+            result.onSuccess { noteId ->
+                scheduleNoteCreate(noteId)
+                setState { state -> state.copy(isAdding = false, added = true) }
+            }.onFailure { message ->
+                setState { state ->
+                    state.copy(isAdding = false, added = false, errorMessage = message)
                 }
-                is ResponseResult.Error -> UIDataState.failed(result.message)
             }
-
-            _addNoteState.emit(state)
         }
     }
 
     private fun scheduleNoteCreate(noteId: String) =
         notyTaskManager.scheduleTask(NotyTask.create(noteId))
+
+    private fun validateNote() {
+        val isValid = NoteValidator.isValidNote(currentState.title, currentState.note)
+        setState { state -> state.copy(showSave = isValid) }
+    }
+
+    fun resetState() {
```

Comment:
This method is called only from simpleapp module. I think it'll be better to mention the reason behind the need for this case.